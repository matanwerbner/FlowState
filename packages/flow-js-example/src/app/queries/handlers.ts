import { rest } from 'msw';
import { QuestionApiResponse } from '../lib/types';

const mockedSuccessResponse: QuestionApiResponse = {
  response_code: 1,
  results: [
    {
      category: 'Entertainment: Video Games',
      type: 'boolean',
      difficulty: 'medium',
      question:
        "The retail disc of Tony Hawk's Pro Skater 5 only came out in North America.",
      correct_answer: 'True',
      incorrect_answers: ['False'],
    },
  ],
};

export const httpHandlers = [
  rest.get('https://opentdb.com/api.php', (req, res, ctx) => {
    return res(ctx.status(200, 'success'), ctx.json(mockedSuccessResponse));
  }),
];
