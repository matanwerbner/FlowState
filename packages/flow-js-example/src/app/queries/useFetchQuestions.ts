import { useQuery } from '@tanstack/react-query';
import { GameContext } from '../lib/types';

export const useFetchQuestions = (
  context: Pick<GameContext, 'category' | 'difficulty'>
) =>
  useQuery({
    queryKey: ['FetchQuestions'],
    networkMode: 'always',
    queryFn: () => {
      const queryParams = new URLSearchParams({
        amount: '10',
        type: 'boolean',
      });
      if (context.category !== '') {
        queryParams.append('category', context.category);
      }
      if (context.difficulty !== '') {
        queryParams.append('difficulty', context.difficulty);
      }

      return fetch(`https://opentdb.com/api.php?${queryParams.toString()}`);
    },
    gcTime: 0,
  });
