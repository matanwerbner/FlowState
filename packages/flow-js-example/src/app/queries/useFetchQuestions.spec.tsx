//Jest tests for useFetchQuestions.ts
import { renderHook, waitFor } from '@testing-library/react';
import { useFetchQuestions } from './useFetchQuestions';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import { setupServer } from 'msw/node';
import { httpHandlers } from './handlers';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const server = setupServer(...httpHandlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

const wrapper = ({ children }: { children: ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe('useFetchQuestions', () => {
  it('should return mocked response for valid data', async () => {
    const { result } = renderHook(
      () => useFetchQuestions({ category: '11', difficulty: 'easy' }),
      {
        wrapper,
      }
    );
    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data?.json()).resolves.toMatchSnapshot();
  });
});
