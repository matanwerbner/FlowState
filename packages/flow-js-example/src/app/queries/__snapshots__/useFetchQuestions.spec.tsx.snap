// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFetchQuestions should return mocked response for valid data 1`] = `
{
  "response_code": 1,
  "results": [
    {
      "category": "Entertainment: Video Games",
      "correct_answer": "True",
      "difficulty": "medium",
      "incorrect_answers": [
        "False",
      ],
      "question": "The retail disc of Tony Hawk's Pro Skater 5 only came out in North America.",
      "type": "boolean",
    },
  ],
}
`;
